ansible-wazuh-agent/tasks/Linux.yml

```diff
@@ -10,60 +10,138 @@
-    /var/ossec/bin/agent-auth
-    -m {{ wazuh_managers.0.address }}
-    -p {{ wazuh_agent_authd.port }}
-    {% if authd_pass is defined %}-P {{ authd_pass }}{% endif %}
-    {% if wazuh_agent_authd.ssl_agent_ca is not none %}
-    -v "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_ca | basename }}"
-    -x "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_cert | basename }}"
-    -k "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_key | basename }}"
-    {% endif %}
-    {% if wazuh_agent_authd.ssl_auto_negotiate == 'yes' %}-a{% endif %}
-  register: agent_auth_output
-  when:
-    - wazuh_agent_authd.enable == true
-    - check_keys.stat.size == 0
-    - wazuh_managers.0.address is not none
-  tags:
-    - config
+- name: Linux | Agent registration via authd
+  block:
 
-- name: Linux | Verify agent registration
-  shell: echo {{ agent_auth_output }} | grep "Valid key created"
-  when:
-    - wazuh_agent_authd.enable == true
-    - check_keys.stat.size == 0
-    - wazuh_managers.0.address is not none
-  tags:
-    - config
+    - name: Retrieving authd Credentials
+      include_vars: authd_pass.yml
+      tags:
+        - config
+        - authd
+
+    - name: Copy CA, SSL key and cert for authd
+      copy:
+        src: "{{ item }}"
+        dest: "/var/ossec/etc/{{ item | basename }}"
+        mode: 0644
+      with_items:
+        - "{{ wazuh_agent_authd.ssl_agent_ca }}"
+        - "{{ wazuh_agent_authd.ssl_agent_cert }}"
+        - "{{ wazuh_agent_authd.ssl_agent_key }}"
+      tags:
+        - config
+        - authd
+      when:
+        - wazuh_agent_authd.ssl_agent_ca is not none
+
+    - name: Linux | Register agent (via authd)
+      shell: >
+        /var/ossec/bin/agent-auth
+        -m {{ wazuh_managers.0.address }}
+        -p {{ wazuh_agent_authd.port }}
+        {% if authd_pass is defined %}-P {{ authd_pass }}{% endif %}
+        {% if wazuh_agent_authd.ssl_agent_ca is not none %}
+        -v "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_ca | basename }}"
+        -x "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_cert | basename }}"
+        -k "/var/ossec/etc/{{ wazuh_agent_authd.ssl_agent_key | basename }}"
+        {% endif %}
+        {% if wazuh_agent_authd.ssl_auto_negotiate == 'yes' %}-a{% endif %}
+      register: agent_auth_output
+      when:
+        - check_keys.stat.size == 0
+        - wazuh_managers.0.address is not none
+      tags:
+        - config
+        - authd
+
+    - name: Linux | Verify agent registration
+      shell: echo {{ agent_auth_output }} | grep "Valid key created"
+      when:
+        - check_keys.stat.size == 0
+        - wazuh_managers.0.address is not none
+      tags:
+        - config
+        - authd
+
+  when: wazuh_agent_authd.enable == true
+
+- name: Linux | Agent registration via rest-API
+  block:
+
+    - name: Retrieving rest-API Credentials
+      include_vars: api_pass.yml
+      tags:
+        - config
+        - api
+
+    - name: Linux | Create the agent key via rest-API
```

Comment:
Let me propose some changes to the API registration tasks:

```
- name: Linux | Create the agent key via rest-API
  uri:
    url: "{{ wazuh_managers.0.api_proto }}://{{ wazuh_managers.0.address }}:{{ wazuh_managers.0.api_port }}/agents/"
    validate_certs: no
    method: POST
    body: {"name":"{{ inventory_hostname }}"}
    body_format: json
    status_code: 200
    headers:
      Content-Type: "application/json"
    user: "{{ wazuh_managers.0.api_user }}"
    password: "{{ api_pass }}"
  register: newagent_api
  changed_when: newagent_api.json.error == 0
  when:
    - check_keys.stat.size == 0
    - wazuh_managers.0.address is not none
  become: no
  tags:
    - config
    - api

- name: Linux | Retieve new agent data via rest-API
  uri:
    url: "{{ wazuh_managers.0.api_proto }}://{{ wazuh_managers.0.address }}:{{ wazuh_managers.0.api_port }}/agents/{{ newagent_api.json.data.id }}"
    validate_certs: no
    method: GET
    return_content: yes
    user: "{{ wazuh_managers.0.api_user }}"
    password: "{{ api_pass }}"
  when:
    - check_keys.stat.size == 0
    - wazuh_managers.0.address is not none
    - newagent_api.json.error == 0
  register: newagentdata_api
  delegate_to: localhost
  become: no
  tags:
    - config
    - api

- name: Linux | Register agent (via rest-API)
  command: /var/ossec/bin/manage_agents
  environment:
    OSSEC_ACTION: i
    OSSEC_AGENT_NAME: '{{ newagentdata_api.json.data.name }}'
    OSSEC_AGENT_IP: '{{ newagentdata_api.json.data.ip }}'
    OSSEC_AGENT_ID: '{{ newagent_api.json.data.id }}'
    OSSEC_AGENT_KEY: '{{ newagent_api.json.data.key }}'
    OSSEC_ACTION_CONFIRMED: y
  register: manage_agents_output
  when:
    - check_keys.stat.size == 0
    - wazuh_managers.0.address is not none
    - newagent_api.changed
  tags:
    - config
    - api
  notify: restart wazuh-agent
```

In this way, we leverage into the API the decision of what IP address should use for the registration of the agent, otherwise, the registration can be useless for NATed agents where the source IP address is deferent than the one used by the server. Thanks in advance!

Regards,

Reply:
@mjflve I can't understand what's the change you proposed :(

Reply:
Hi @angystardust,

The main idea is to let the Wazuh API decide the IP address used for the registration, in your approach you delegate that option to the registration request via the IP address detected in `ansible_default_ipv4.address` making the request from the Ansible control server, and that will work in most of the cases. What I suggest is to run the request from the Ansible client without the `ip` argument on the request, then the Wazuh API will use the origin address as the agent registration IP address, this could be helpful for agents that resided behind a NATed network where the traffic is masqueraded over a gateway, this in deep because the manager ensures that the agent IP address should be the same that origin the connection.

I hope this could clarify my approach, Thanks!

Best Regards,

Reply:
Hi @mjflve sorry but I didn't had time to reply and modify the code. Feel free to implement your suggestion so we can merge my PR in the  main repository.
Thanks